File: 1654-minimum-jumps-to-reach-home/1654-minimum-jumps-to-reach-home.cpp
class Solution {
 unordered_map<int, int> um;
    int dfs(int a, int b, int x, int val, int pn, vector<vector<int>> &dp)
    {
        if(val == x) return 0;
        if(um.count(val)) return 6000;
        if(val < 0 || val > 6000) return 6000;
        if(dp[val][pn] != -1)
            return dp[val][pn];
        dp[val][pn] = 1 + dfs(a,b,x,val+a, 0, dp);
        if(pn == 0)
            dp[val][pn] = min(dp[val][pn], 1+dfs(a, b, x, val-b, 1,dp));
        return dp[val][pn];
    }
public:
    int minimumJumps(vector<int>& forbidden, int a, int b, int x) {
        vector<vector<int>> dp(6001, vector<int> (2, -1));
        for(int &num : forbidden)
        {
            if(num == x) return -1;
            um[num]++;
        }
        int val = dfs(a, b, x, 0, 0, dp);
        if(val >= 6000) return -1;
        return val;
    }
};

File: 1654-minimum-jumps-to-reach-home/1654-minimum-jumps-to-reach-home.java
class Solution {
   public int minimumJumps(int[] forbidden, int a, int b, int x) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a1,a2)-> a1[0] - a2[0]);
        pq.offer(new int[]{0,0,0});//step, current index, direction(0 is back, 1 is forward)
        Set<Integer> forbit = new HashSet<>();
        Set<String> visited = new HashSet<>();
        int maxLimit = 2000 + 2 * b;
        for(int num : forbidden){
            forbit.add(num);
            maxLimit = Math.max(maxLimit, num + 2 * b);
        }
        while(!pq.isEmpty()){
            int[] node = pq.poll();
            int step = node[0];
            int idx = node[1];
            int dir = node[2];
            if(idx == x) return step;
			//try jump forward
            if(idx+a < maxLimit && !forbit.contains(idx+a) && !visited.contains(idx+a+","+0)){
                visited.add(idx+a+","+0);
                pq.offer(new int[]{step+1, idx+a, 0});
            }
			//try jump back
            if(idx-b >= 0 && !forbit.contains(idx-b) && !visited.contains(idx-b+","+1) && dir != 1){
                visited.add(idx-b+","+1);
                pq.offer(new int[]{step+1, idx-b, 1});
            }
        }
        return -1;
    }
}